### DISCUSSION TITLE
Hosting and maintaining a deno web server in production

### DISCUSSION THREAD

**Discussion by domo84** [denoland/deno#14010] [Q&A] [answered]
This question is about the surrounding environment and tools that is needed to host and maintain a stable deno web server in production.

I have a web server written in JS that I run with `deno run --allow-net webserver.js` inside a `screen` session on Ubuntu. Web server is exposed at `localhost:8000`. I then have `apache2` acting as a proxy, mapping incoming connections from the outside via a domain name to the web server. This is basically what every guide I found says (perhaps minus `screen` and more in favour of `nginx`). It works.

Problem is deno crashes every now and then because of bugs in my code and with the current setup I receive no alert and I would like deno to restart the web server after a crash. And it isn't reboot friendly either.

I can think of multiple ways of solving the problems, like `systemd` and `syslog`, but I feel like I'm missing important pieces of the puzzle that already have been solved many times over.

How does one host, manage and monitor these type of applications?

Useful tips and pointers in the right direction is greatly appreciated.

**Comment by elycheikhsmail** [comment, marked as answer]
pm2 is js (node package) can used in your situation
or use managed service like herokue or  deno deploy if they satisfy your need.
But I think we need official guid and doc for self hosting deno app in production( in vps ) in deno manual

**Reply by domo84** [reply]
I didn't know about PM2 until now. It indeed looks like it can be used in my situation. I will try it out. Thank you!

An official deno guide on this topic would be very helpful.

And by the way I tried deno deploy but didn't work in my case as I am using the File API extensively which is not supported.
